Add lemma divzmul_divdiv and strenghten/proof modz_pow2_div.

MM45 opened the pull request:
Added one lemma (`divzmul_divdiv`) stating that dividing by `a * b` is equivalent to first dividing by `a` and then dividing by `b` ( 'dividing' refers to 'perform euclidean division and take the quotient' here). Using this lemma, proved lemma `modz_pow2_div` that was `admitted` until now. Also, turned out one of the preconditions in `modz_pow2_div` (`0 <= i`) was not necessary, so I removed it.

strub commented on theories/algebra/IntDiv.ec:
```diff
@@ -864,9 +864,33 @@ move=> lt_pn; rewrite {1}(divz_eq i (k^n)); rewrite -addrA; congr.
-lemma nosmt modz_pow2_div n p i: 0 <= i => 0 <= p <= n =>
-  (i %% 2^n) %/ 2^p = (i %/ 2^p) %% 2^(n-p).
-proof. admitted.
+lemma nosmt divzmul_divdiv i a b :  
```
Rename to `divzMr`.

fdupress commented on theories/algebra/IntDiv.ec:
```diff
@@ -864,9 +864,37 @@ move=> lt_pn; rewrite {1}(divz_eq i (k^n)); rewrite -addrA; congr.
+  0 <= a => 0 <= b => i %/ (a * b) = i %/ a %/ b.
+proof.
+move=> ge0_a ge0_b.
+case (a = 0); first by smt(div0z). 
```
It's possibly slightly better for readability to introduce `gt0_a` here, rather than after the second case.

MM45 replied:
I assume you mean `neq0_a` instead of `gt0_a`?

fdupress replied:
Yes, sorry.

fdupress commented on theories/algebra/IntDiv.ec:
```diff
@@ -864,9 +864,37 @@ move=> lt_pn; rewrite {1}(divz_eq i (k^n)); rewrite -addrA; congr.
+  0 <= a => 0 <= b => i %/ (a * b) = i %/ a %/ b.
+proof.
+move=> ge0_a ge0_b.
+case (a = 0); first by smt(div0z). 
```
I just can't see how `smt(div0z)` is solving this. Are we missing core lemmas about `%%` that are true in the `SMTLIB` model? (Don't answer that, simply tell me if you have a non-smt proof.)

fdupress replied:
This one truly has me worried: Alt-Ergo has messed up Euclidean division in the past and we should understand what the SMT solvers are doing.

fdupress replied:
So this is `by rewrite divz0 div0z.`

The fact that `smt(divz0)` finds this proof means that one of the following is true:
1) SMTLIB specifies division by 0; or
2) `divz0` is a trivial (for SMT) consequence of the definition of `edivz`.

If neither is true, then we should dig a bit deeper.

MM45 replied:
Yes, I also noticed this oddity (however, instead of being critical like you, I simply assumed it should be alright) as I first did this with regular rewriting. However, to shorten proofs, I often replace annoying rewrites with only relatively simple or rudimentary (e.g. associativity, commutativity, multiplication by 0, etc.) by an SMT call with the appropriate lemmas given. I find the 'appropriate' lemmas by simply passing all the lemmas I used in the rewrite (that are slightly non-trivial) and then trimming them down to see what smt minimally needs. This led to the `smt(div0z)` in this case, even though, as you can see in the current code (and as you already noted), the rewrite indeed also requires `divz0`. Nevertheless, for now, I removed all SMT calls just to be sure.

strub replied:
For ring/field equalities, you can use the `ring` / `field` tactic.

fdupress commented on theories/algebra/IntDiv.ec:
```diff
@@ -864,9 +864,37 @@ move=> lt_pn; rewrite {1}(divz_eq i (k^n)); rewrite -addrA; congr.
+have ->: i %/ a %/ b * ab + (i %/ a %% b * a + i %% a) = i.
++ by rewrite /ab (mulrC a) mulrA addrA -mulrDl -eqia_b -eqi_a.
+rewrite -eqi_ab rngi_ab andaE /=.
+have -> //: 0 <= i %/ a %% b * a + i %% a /\ i %/ a %% b * a + i %% a < `|ab|.
```
Here, you can avoid the `have -> //: <some long-winded formula that's already in the goal>` with `move=> /(_ _) 2:-> //.`

`move=> /(_ _)` applies the top assumption to an anonymous proof of its first premise, requesting that proof as the first subgoal, and leaving the second subgoal with the top assumption replaced with the conclusion of that application.

MM45 replied:
Wow, I didn't know that; that's a cool feature. And, apparently, it also works with even more underscores if your top assumption is a series of more implications; really nice!

strub replied:
You don't even need this here. You can simply `move=> -> //`. EasyCrypt will do the required applications until it finds an equality to rewrite.

MM45 replied:
Hmm, maybe I am doing it wrong, but I am having trouble achieving the same with `move=> -> //`. I think that it is maybe due to the conclusion of the top assumption being a conjunction of which I only need to rewrite with the left operand? That is, at the point where I use `move=> /(_ _ _ _) //` now, the goal is:
```
(i %/ ab * ab + i %% ab = i %/ a %/ b * ab + (i %/ a %% b * a + i %% a) =>
 0 <= i %% ab && i %% ab < `|ab| =>
 0 <= i %/ a %% b * a + i %% a && i %/ a %% b * a + i %% a < `|ab| =>
 i %/ ab = i %/ a %/ b /\ i %% ab = i %/ a %% b * a + i %% a) =>
i %/ ab = i %/ a %/ b
```
Using `move => ->` gives an error here. Even (for testing purposes) changing the `/\` to a `=>` does not allow me to use `move => -> //`. However, when the thing I want to rewrite is the actual (full) conclusion of the top assumption, it does work. I.e., if I change the above to the following, `move=> -> //` works:
```
(i %/ ab * ab + i %% ab = i %/ a %/ b * ab + (i %/ a %% b * a + i %% a) =>
 0 <= i %% ab && i %% ab < `|ab| =>
 0 <= i %/ a %% b * a + i %% a && i %/ a %% b * a + i %% a < `|ab| =>
 i %/ ab = i %/ a %/ b) =>
i %/ ab = i %/ a %/ b
```

fdupress replied:
Your understanding is correct: `rewrite` wants a conclusion that is an equality (or an equivalence), and instead finds a conjunction. It then interprets that conjunction as being equal to `true` and tries to rewrite that, but that also fails.

@strub perhaps we could think of an intro pattern that rewrites conjunctions of equalities by (tentatively) rewriting each of the conjuncts?

strub replied:
Yes, that would mimic the behavior of ssr when it encounters a tuple of equalities.

fdupress commented on theories/algebra/IntDiv.ec:
```diff
@@ -864,9 +864,37 @@ move=> lt_pn; rewrite {1}(divz_eq i (k^n)); rewrite -addrA; congr.
+split; first by rewrite addz_ge0 1:mulr_ge0 1,3:modz_ge0.
+rewrite ger0_norm 1:mulr_ge0 // (ltr_le_trans ((b - 1) * a + a) _ ab).
++ by rewrite ler_lt_add 1:ler_pmul 1:modz_ge0; smt(ltz_pmod).
++ by rewrite mulzDl mulNr -addzA addNz mulzC.
```
De-indent one level.

fdupress commented on theories/algebra/IntDiv.ec:
```diff
@@ -864,9 +864,37 @@ move=> lt_pn; rewrite {1}(divz_eq i (k^n)); rewrite -addrA; congr.
+rewrite -eqi_ab rngi_ab andaE /=.
+have -> //: 0 <= i %/ a %% b * a + i %% a /\ i %/ a %% b * a + i %% a < `|ab|.
+split; first by rewrite addz_ge0 1:mulr_ge0 1,3:modz_ge0.
+rewrite ger0_norm 1:mulr_ge0 // (ltr_le_trans ((b - 1) * a + a) _ ab).
```
Would the `rewrite (ltr_le_trans ...)` allow slightly inference as an `apply: (ltr_le_trans ((b - 1) * a + a))`?

MM45 commented:
@fdupress, I just implemented your changes (and rebased such that it is only a single commit). Hope I did everything correctly. Let me know if anything is still out of order.
Also, I was thinking, should `divzMr` and `divzMl` be placed a bit higher in the file, closer to other similar lemmas?

strub commented:
> Also, I was thinking, should `divzMr` and `divzMl` be placed a bit higher in the file, closer to other similar lemmas?

Yes, this seems a good idea to move them up.

MM45 commented:
> > Also, I was thinking, should `divzMr` and `divzMl` be placed a bit higher in the file, closer to other similar lemmas?
> 
> Yes, this seems a good idea to move them up.

Should I do this or do you want to do this during the merging process?

strub commented:
> Should I do this or do you want to do this during the merging process?

I'll take care of it and have a look at this `divz0` / `div0z` discussion.
